File: src/instructions/tools/gather-context/gather-context.py
#!/usr/bin/env python3
"""gather-context — orchestrate context-gathering tools for a set of keywords.

Currently orchestrates:
  1. gather-context-from-memory — search memory vault, return full file bodies

More tools will be added here over time.

Usage:
  python3 gather-context.py planning workflow
  python3 gather-context.py planning --format json
  python3 gather-context.py planning --markdown
"""
import argparse
import json
import os
import subprocess
import sys
from pathlib import Path


def find_telamon_root(start: Path) -> Path:
    p = start.resolve()
    for _ in range(10):
        if (p / "src" / "instructions" / "tools").is_dir():
            return p
        p = p.parent
    return start.resolve()


SCRIPT_DIR = Path(__file__).resolve().parent
TELAMON_ROOT = find_telamon_root(SCRIPT_DIR)
TOOLS_DIR = TELAMON_ROOT / "src" / "instructions" / "tools"


def run_tool(script: Path, extra_args: list[str]) -> tuple[str, int]:
    env = {**os.environ, "XDG_CACHE_HOME": str(TELAMON_ROOT / "storage"), "QMD_LLAMA_GPU": "true"}
    proc = subprocess.run(["python3", str(script)] + extra_args, capture_output=True, text=True, env=env)
    return proc.stdout, proc.returncode


def resolve_collection(project_root: Path) -> str:
    config = project_root / ".ai" / "telamon" / "telamon.jsonc"
    try:
        return json.loads(config.read_text()).get("project_name", "telamon")
    except Exception:
        return "telamon"


def main() -> None:
    parser = argparse.ArgumentParser(description="gather-context: orchestrate context-gathering tools")
    parser.add_argument("keywords", nargs="*", help="Keywords to search for (positional)")
    parser.add_argument("--format", choices=["markdown", "json"], default="markdown")
    parser.add_argument("--markdown", action="store_true")
    parser.add_argument("--json", action="store_true", dest="json_flag")
    parser.add_argument("--collection", default="")
    parser.add_argument("--max-results", type=int, default=5)
    args = parser.parse_args()

    fmt = args.format
    if args.markdown:
        fmt = "markdown"
    if args.json_flag:
        fmt = "json"

    keywords: list[str] = list(args.keywords)
    if not keywords:
        print("Error: provide at least one keyword.", file=sys.stderr)
        sys.exit(1)

    project_root = Path(os.getcwd())
    collection = args.collection or resolve_collection(project_root)

    # ── 1. gather-context-from-memory ────────────────────────────────────────
    memory_script = TOOLS_DIR / "gather-context-from-memory" / "gather-context-from-memory.py"
    memory_args = [
        "--format", fmt,
        "--collection", collection,
        "--max-results", str(args.max_results),
    ]
    for kw in keywords:
        memory_args += ["--query", kw]

    memory_out, memory_code = run_tool(memory_script, memory_args)

    # ── Output ────────────────────────────────────────────────────────────────
    if fmt == "json":
        result: dict = {}
        if memory_code == 0:
            try:
                result.update(json.loads(memory_out.strip()))
            except json.JSONDecodeError:
                result["memory_error"] = memory_out.strip()
        else:
            result["memory_error"] = memory_out.strip()
        print(json.dumps(result, indent=2))
    else:
        parts = []
        if memory_code == 0 and memory_out.strip():
            parts.append(memory_out.strip())
        elif memory_code != 0:
            parts.append(f"_Memory search failed: {memory_out.strip()}_")
        print("\n\n".join(parts))


if __name__ == "__main__":
    main()

File: src/instructions/tools/gather-context/gather-context.ts
import { tool } from "@opencode-ai/plugin"
import path from "path"
import fs from "fs"

/**
 * gather-context — orchestrate context-gathering tools for a set of keywords.
 *
 * Currently orchestrates:
 *   1. gather-context-from-memory — search memory vault, return full file bodies
 *
 * More tools will be added over time.
 *
 * Usage:
 *   gather-context({ keywords: ["planning", "workflow"] })
 *   gather-context({ keywords: ["memory"], format: "markdown" })
 *
 * Delegates to the colocated Python script (gather-context.py).
 *
 * Wiring:
 *   - This file lives at <telamon-root>/src/instructions/tools/gather-context/gather-context.ts
 *   - init.sh creates a flat symlink at <project>/.opencode/tools/gather-context.ts
 *   - `@opencode-ai/plugin` is at src/instructions/tools/node_modules/
 */

function resolveProjectCollection(): string {
  const configPath = path.join(process.cwd(), ".ai", "telamon", "telamon.jsonc")
  try {
    const raw = fs.readFileSync(configPath, "utf8")
    const config = JSON.parse(raw)
    if (config.project_name) return config.project_name
  } catch {
    // Fall through to default
  }
  return "telamon"
}

export default tool({
  description:
    "Gather context for a set of keywords by orchestrating multiple context sources. Currently searches the memory vault (notes, decisions, patterns). More sources will be added over time. Use at the start of any non-trivial session to prime context about a topic.",
  args: {
    keywords: tool.schema
      .array(tool.schema.string())
      .describe(
        "One or more keywords or phrases describing the topic. E.g. ['planning', 'workflow'] or ['memory', 'skill'].",
      ),
    format: tool.schema
      .enum(["json", "markdown"])
      .optional()
      .default("json")
      .describe("Output format: 'json' (default) or 'markdown' (human-readable)"),
    max_results: tool.schema
      .number()
      .optional()
      .default(5)
      .describe("Maximum number of memory vault results (default: 5)"),
    collection: tool.schema
      .string()
      .optional()
      .describe("QMD collection name (default: auto-detected from .ai/telamon/telamon.jsonc project_name)"),
  },
  async execute(args) {
    const script = path.join(import.meta.dir, "gather-context.py")
    const fmt = args.format ?? "json"
    const collection = args.collection ?? resolveProjectCollection()

    const cmd = [
      "python3",
      script,
      "--format", fmt,
      "--collection", collection,
      "--max-results", String(args.max_results ?? 5),
      ...args.keywords,
    ]

    const env = {
      ...process.env,
      XDG_CACHE_HOME: path.join(import.meta.dir, "../../../../storage"),
      QMD_LLAMA_GPU: "true",
    }

    const proc = Bun.spawn(cmd, { stdio: ["ignore", "pipe", "pipe"], env })
    const stdout = await new Response(proc.stdout).text()
    const stderr = await new Response(proc.stderr).text()
    const exitCode = await proc.exited

    if (exitCode !== 0) {
      return `gather-context failed (exit ${exitCode})\n${stderr.trim() || stdout.trim() || "(no output)"}`
    }

    if (fmt === "json") {
      try {
        return JSON.parse(stdout.trim())
      } catch {
        return stdout.trim()
      }
    }

    return stdout.trim()
  },
})
